src/main/scala/calc/Compiler.scala

```diff
@@ -61,12 +64,34 @@ object Compiler {
     ir.Hashers.hashClassDef(classDef)
   }
 
+
+
+  def findCapture(tree: Tree): Set[String] = {
+    tree match {
+      case _:Literal => Set()
+      case BinaryOp(_, l, r) => findCapture(l) ++ findCapture(r)
+      case Let(i, v, body) => findCapture(body) ++ findCapture(v) - i.name
+      case Ident(n) => Set(n)
+      case If(cond, thenp, elsep) => findCapture(cond) ++ findCapture(thenp) ++ findCapture(elsep)
+      case Closure(idents, body) => findCapture(body) -- idents.toSet.flatMap((x:Tree) => findCapture(x))
+      case Call(f, args) => findCapture(f) ++ args.toSet.flatMap((x:Tree) => findCapture(x))
+    }
+  }
+
+  def compileClosure(tree: Closure, ident: String)(implicit env:Map[String, irt.Tree],typeEnv:Map[String, irtpe.Type]): irt.Tree ={
+    val captureSet = findCapture(tree) - ident
+    val captureParam = captureSet.map(s => irt.ParamDef(irt.Ident(s)(tree.pos), typeEnv(s), false, false)(tree.pos)).toList
+    val params = tree.params.map(s => irt.ParamDef(irt.Ident(s.name)(tree.pos), irtpe.AnyType, false, false)(tree.pos))
```

Comment:
So, here, you define parameters of type `any`. And then, *every* time a parameter is used, you cast it down to a `double` with `Unbox(x, 'D')`. In the typical spirit of erasure for Java and Scala, it would be better to downcast them *once*, upon entry in the function.

That means parameters will need a name different than `s.name` (a generated one), since that name will be used for a local `VarDef` of type `Double`, which will store the result of `Unbox`. Then that local variable can be used in the body of the method.